pallets/xvm/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! # XVM pallet
+//!
+//! ## Overview
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Function
+//!
+//!
+//! ### Other
+//!
+//!
+
+use frame_support::{pallet_prelude::*, traits::ConstU32, BoundedVec};
+use pallet_evm::GasWeightMapping;
+use parity_scale_codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use sp_core::U256;
+use sp_runtime::{traits::StaticLookup, RuntimeDebug};
+use sp_std::{marker::PhantomData, prelude::*, result::Result};
+
+use astar_primitives::ethereum_checked::{
+    AccountMapping, CheckedEthereumTransact, CheckedEthereumTx, MAX_ETHEREUM_TX_INPUT_SIZE,
+};
+
+pub use pallet::*;
+
+// TODO: move types and traits to `astar-primitives`.
+
+/// XVM call info on success.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct CallInfo {
+    /// Output of the call.
+    pub output: Vec<u8>,
+    /// Actual used weight.
+    pub used_weight: Weight,
+}
```

Comment:
Besides output and weight we can also forward the evm logs we get from evm tx appy via call_info to users.
Can be useful for debugging contract reverts.

Reply:
There is an open ticket for cross-VM contract reverts. I'll handle the error info propagation when working on it.